### DISCUSSION TITLE
How to save song title and artist of songs from streaming radio to a file

### DISCUSSION THREAD

**Discussion by dln949** [MusicPlayerDaemon/MPD#1719] [Q&A]
Using mpd I play a radio stream.  If possible, I'd like mpd to record in a file the song's artist and title to a file (append, not overwrite).  

Is that possible?  

How to do that?

**Comment by dcabri** [comment]
It is not impossible, but it looks like MPD refuses to save it automatically into some kind of file, or even simply logging.
So, I created a script to do just that. It's a little bit using resources, but at least it keeps a single connection open (much in stead of doing multiple mpc calls).

I would very much love mod would save these strings (if so configured) somewhere, in whatever format (!)

Here goes my bash script:



METADATA_LOG="/mnt/work/log/mpd_metadata.log"
PREVIOUS_songtitle=""
PREVIOUS_stationname=""
getinfo() {
    while true; do
        sleep 1
        echo "currentsong"
    done | stdbuf -oL nc 127.0.0.1 6600

}
while true ; do
getinfo | while read line ; do
  case "$line" in
    Title:*)
      songtitle=${line#*: }
      if [ "$songtitle" != "$PREVIOUS_songtitle" ]; then
           echo "$(date +%s) $songtitle" >> "$METADATA_LOG"
           PREVIOUS_songtitle=$songtitle
      fi
      ;;
    Name:*)
      stationname=${line#*: }
      if [ "$stationname" != "$PREVIOUS_stationname" ]; then
           echo "$(date +%s) STATION: $stationname" >> "$METADATA_LOG"
           PREVIOUS_stationname=$stationname
      fi
      ;;
  esac
done

**Comment by MaxKellermann** [comment]
https://mpd.readthedocs.io/en/stable/plugins.html#recorder

**Comment by dcabri** [comment]
Thank you for this. But this saves the audio stream into a file.
MPD is playing audio (and in my case piping to camilladsp).
In such case, we would think it would be nice thing to have the metadata in some file,
next to the audio stream...
I don't think the link provided allows for this.
But maybe I am wrong, and in such case, please feel free to enlighten me.
Thank you for your time!

**Reply by MaxKellermann** [reply]
MPD saves metadata to the recorded music files.

**Reply by dcabri** [reply]
So... can a future mpd maybe save metadata (on request) to another (text) file too?
use case is updating on-screen information when using a display, which would not need to query the mpd daemon endlessly....

**Reply by MaxKellermann** [reply]
I don't see a point in that, and you havn't explained this. But if you want to suggest a new feature, write a feature request ticket and write a proper explanation. Or just hack the code and maybe submit it as a PR.

**Reply by dcabri** [reply]
I'd like to propose an improvement related to handling metadata updates when playing an internet radio stream in MPD.

Currently, when MPD receives metadata updates, I can query them using the mpc current command. However, to keep track of what is currently playing, I need to repeatedly run this command at regular intervals. This method, while effective, is not efficient in terms of CPU usage.

The purpose of accessing this metadata is to make it available for other processes, such as storing information about played songs or updating a display. Therefore, I would like to propose a more efficient mechanism for handling this data.

Specifically, I suggest that when an internet radio stream is played, MPD should:

Write the incoming metadata to a text file, or
Launch a script each time new metadata arrives.
Ideally, either of these options would be configurable through the mpd.conf file, allowing users to enable or disable them and adjust parameters as needed.

I believe this enhancement would provide a significant benefit to anyone integrating MPD with other tools or automation processes.

**Reply by MaxKellermann** [reply]
I think you misunderstand @borine's code snippet. You do *not* query MPD in "regular intervals". You only query MPD after MPD has notified you that something has changed (which @borine implemented with `mpc idleloop`). Therefore your assumptions about CPU usage are bogus.
Your suggested feature is bad because:
- it doesn't actually save any CPU cycles
- works only if MPD and your application are running on the same machine
- invents a new non-standard file format
- nobody supports this file format, but there are many applications which support the MPD protocol

**Reply by dcabri** [reply]
I'm sorry, but
 "You only query MPD after MPD has notified you that something has changed"
 Is not valid for internet radio.
 When one opens an internet radio stream, after that nothing changes anymore.
 Except for the metadata, but there is no notification for that...

**Reply by dcabri** [reply]
OMG
I take back my previous comments.
I am SO sorry.
Indeed, mpc idleloop
generates output:
playlist
player
when the metadata changes.

My deepest apologies!

Thank you so much for pointing that out!

**Comment by borine** [comment]
Here's a simple shell script using mpc to append song metadata to a file. Experiment with different format strings  to get the specific metadata items that interest you (see `man mpc`)
```shell
while true ; do
        mpc --wait --format "[[%artist% - ]%title%]|[%name%]" current >>$filename
done
```

**Reply by dcabri** [reply]
This is cool for playlists, but for an internet radio url stream, it does not update the metadata information.
So, would be nice to have a possibility to receive metadata updates in real time in some way.
btw, for me, I did not want to write all metadata into the audio file, 
rather for updating the on-screen information, I would not have to use nc or mpc with repeatedly sending same commands...

**Reply by borine** [reply]
> would be nice to have a possibility to receive metadata updates in real time in some way

The mpd API certainly permits this. To do it without repeatedly calling a one-shot application you would have to search around to see what is available, or else write your own client. For desktop use I think there are mpd MPRIS clients available.

Calling mpc is not a very heavy load on the system, I think, especially if the system is powerful enough to run a desktop. If the first script above does not pick up radio metadata updates, try:

```shell
mpc idleloop playlist | while read ; do
	mpc --format "[[%artist% - ]%title%]|[%name%]" current >>$filename
done
```

**Reply by MaxKellermann** [reply]
Maybe some perspective on that: switching tabs on your browser probably consumes a thousand times (or ten thousand times or more) more CPU cycles than this `mpc` loop. Of course nobody wants to waste CPU cycles, but the same people use Electron-based applications and do everything in a gigabytes of RAM consuming browser. MPD is a time traveller from a distant past where low resource consumption was taken seriously.

**Reply by dcabri** [reply]
I understand, thank you for that.
It is a GREAT piece of software, thank you, and thanks to the developers for creating it!
I use it on raspberry pi, so I'm cpu cycle conscious :)
